lib/tables.dart

```diff
@@ -122,66 +176,203 @@ class Table<T> {
+    else
+      _rebuildStaticTable();
+  }
 
-    // Re-build the table.
-    final List<Element> rowElements = <Element>[];
+  void _rebuildVirtualTable() {
+    // When we modify the height of the spacer above the viewport, Chrome automatically
+    // adjusts the scrollTop to compensate because it's trying to keep the same content
+    // visible to the user. However, since we're overwriting the rows contents (shifting
+    // the data in the rows) we want to retain the original scroll position so must stash
+    // it and reset it later.
+    final int originalScrollTop = element.scrollTop;
+
+    int firstRenderedRowInclusive = 0;
+    int lastRenderedRowExclusive = rows?.length ?? 0;
+
+    // Keep track of the table row we're inserting so that we can re-use rows
+    // if they already exist in the DOM.
+    int currentRowIndex = 0;
+
+    // Calculate the subset of rows to render based on scroll position.
+    final int firstVisibleRow =
+        ((element.scrollTop - _thead.offsetHeight) / rowHeight).floor();
+    final int numVisibleRows = (element.offsetHeight / rowHeight).ceil();
+    firstRenderedRowInclusive = firstVisibleRow.clamp(0, rows?.length ?? 0);
+    // Calculate the last rendered row. +2 is for:
+    //   1) because it's exclusive so needs to be one higher
+    //   2) because we need to render the extra partially-visible row
+    lastRenderedRowExclusive = (firstRenderedRowInclusive + numVisibleRows + 2)
+        .clamp(0, rows?.length ?? 0);
+
+    // Add a spacer row to fill up the content off-screen.
+    final double spacerBeforeHeight = firstRenderedRowInclusive * rowHeight;
+    _spacerBeforeVisibleRows.height = '${spacerBeforeHeight}px';
+    _spacerBeforeVisibleRows.display = spacerBeforeHeight == 0 ? 'none' : null;
+
+    // If the spacer row isn't already at the start of the list, add it.
+    if (_tbody.element.children.isEmpty ||
+        _tbody.element.children.first != _spacerBeforeVisibleRows.element) {
+      _tbody.element.children.insert(0, _spacerBeforeVisibleRows.element);
+    }
+    currentRowIndex++;
+
+    // Remove the last row if it's the spacer - we'll re-add it later if required
+    // but it simplifies things if we can just append any new rows we need to create
+    // to the end.
+    if (_tbody.element.children.isNotEmpty &&
+        _tbody.element.children.last == _spacerAfterVisibleRows.element) {
+      _tbody.element.children.removeLast();
+    }
 
-    for (T row in rows) {
-      final CoreElement tableRow = tr();
+    currentRowIndex = _buildTableRows(
+        firstRenderedRowInclusive: firstRenderedRowInclusive,
+        lastRenderedRowExclusive: lastRenderedRowExclusive,
+        currentRowIndex: currentRowIndex);
+
+    // Remove any additional rows that we had left that we didn't reuse above.
+    if (currentRowIndex > 0 &&
+        currentRowIndex < _tbody.element.children.length) {
+      // TODO: Why does removeAt(i) work, but not removeRange (UnimplementedError)?
```

Comment:
It's a long-standing issue with dart:html. I'd change this todo: into just a regular comment mentioning why we don't use removeRange (cc @jacob314).

Reply:
Ok cool. I think this code is actually buggy, since it's removing an item but still incrementing the count (so it'll skip an item for each one it removes). I'll fix!